KEP-4601: authorize with field and label selectors (#4600)
kubernetes/enhancements: keps/sig-auth/4601-authorize-with-selectors/README.md

- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

The authorization attributes will be extended to include field selectors and label selectors from

> Get, List, Watch, and Deletecollection.

Reviewer: ```suggestion
List, Watch, and Deletecollection.
```

This will allow authorizers to use these selectors when making an authorization decision.

## Motivation

Security for per-node workloads could be improved by exposing field and label selectors to authorizers.
Adding them as authorization attributes allows the development of new kinds of authorizers that
leverage this information to provide security.
In particular, it enables out-of-tree authorizers to experiment with ways to express restrictions based on field and label selectors.

### Goals

* Add field and label selectors to authorization attributes for Get, List, Watch, and DeleteCollection verbs.
* Add field and label selectors to webhook authorization types.
* Add field and label selectors to SelfSubjectAccessReview (SSAR) and SubjectAccessReview (SAR).
  <<[UNRESOLVED deads2k, liggitt, aws-Micah: you guys volunteering to deliver this bit? ]>>
* Update node authorizer to restrict on nodeName field selector.
  <<[/UNRESOLVED]>>

### Non-Goals

> * Create a generic in-tree authorizer that manages field or label selectors.

Reviewer: A couple more non-goals to make explicit:
* I don't think we need to expand the audit surface area, since requestURI is already included
* I also don't think we need to expand the admission surface area (admission.Attributes, AdmissionReview, CEL `authorizer` implementation available to admission) since admission verbs don't support field/label selectors

Author: > * since admission verbs don't support field/label selectors

There's no admission on DeleteCollection?

Reviewer: > There's no admission on DeleteCollection?

admission gets invoked N times, once for every object actually deleted

## Proposal

List, Watch, and DeleteCollection requests directly have field and label selector options.

> A single-item List or Watch request is remapped to the Get verb, so Get requests can have field and label selectors.

Reviewer 2: I can see a huge value for being able to authorize based on `metadata.name` field (so field selectors).

What is the usecase for the labelSelector? Is it effectively added only for downstream extensibility, or do you see any upstream usecase too? If it's the former, it would be useful to add it explicitly to the KEP.

Author: > Is it effectively added only for downstream extensibility, or do you see any upstream usecase too?

Downstream extensibility of APIs by non-API owners.

> A single-item List or Watch request is remapped to the Get verb, so Get requests can have field and label selectors.

Reviewer: > A single-item List or Watch request is remapped to the Get verb, so Get requests can have field and label selectors.

That's not what happens today, right? A list or watch request with a metadata.name field selector populates resourceName in request info, but does not change the verb.

### Authorization Attributes changes

The authorization attributes have easy access to the query parameter field and label selectors.
To avoid confusion, field and label selectors will not be included in authorization attributes for verbs where the field

> selector has no semantic meaning.

Reviewer: what will kube-apiserver do if presented with a SubjectAccessReview that sets the selectors in combination with verbs other than list/watch/deletecollection?

Author: > what will kube-apiserver do if presented with a SubjectAccessReview that sets the selectors in combination with verbs other than list/watch/deletecollection?

I would like for webhooks to write code to handle field and label selection for any verb.  The kube-apiserver must only include field and label selectors in authorization attributes when they are honored.

This means that the SAR to the kube-apiserver must remove them from verbs where they aren't honored. But if the kube-apiserver grows the ability to honor them on say an update, then a webhook should also honor the restriction.  Ideally without modification.

Reviewer: > if the kube-apiserver grows the ability to honor them on say an update, then a webhook should also honor the restriction. Ideally without modification.

That seems ideal, though to write code to honor this now, the semantics of what a selector on a write means probably have to be defined (must the selector match the incoming object, the existing object, either, or both?)

It's somewhat weird to define semantics for behavior which doesn't yet exist... the only prior art I can think of are `objectSelector` on admission. That takes the most conservative stance, if either old object or new object matches, the webhook is called:

> objectSelector is evaluated against both the oldObject and newObject, and is considered to match if either object matches the selector.

The most conservative stance on writes would be that both the old and new object must fall within the selector, I guess.

Reviewer: > The kube-apiserver must only include field and label selectors in authorization attributes when they are honored.
>
> This means that the SAR to the kube-apiserver must remove them from verbs where they aren't honored.

So the SAR caller would be responsible for only populating the selector fields when they would be used, kube-apiserver wouldn't strip them out of presented SARs?

And SARs kube-apiserver generates internally would only include the selectors on verbs where they are used? I guess kube-apiserver will assume aggregated servers will support exactly what kube-apiserver does for list / watch selectors.

Author: > And SARs kube-apiserver generates internally would only include the selectors on verbs where they are used? I guess kube-apiserver will assume aggregated servers will support exactly what kube-apiserver does for list / watch selectors.

And the aggregated apiservers are explicitly told to authorize all requests themselves, so the aggregated APIs are well covered.

> So the SAR caller would be responsible for only populating the selector fields when they would be used, kube-apiserver wouldn't strip them out of presented SARs?

I rephrased on the push.

client -> kube-apiserver SAR endpoint: if the kube-apiserver would not honor the selectors, they are removed prior to calling the authorizer.

client -> webhook: the client is responsible for describing what they mean

kube-apiserver -> webhook: the kube-apiserver will send selectors for all verbs where the selector is honored.  I'll  add a section describing potential future semantics for field selector on Create, Update, and Delete.

In practice this means that (for now), only Get, List, Watch, and DeleteCollection have field and label selectors.
Any authorizer that gets an error from `ParseFieldSelector` or `ParseLabelSelector` may attempt to authorize with the 
`GetFieldSelector` or `GetLabelSelector` or attempt to authorize without field or label selectors since that will authorize
using a wider permission (field and label selectors can only reduce access).

```go
type Attributes interface {
	// GetFieldSelector returns the query parameter value of the field selector
	// Remember that only List, Watch, and DeleteCollection (currently) provide 
	// Remember that field selector formats vary based on the version of the API being used!
	GetFieldSelector() string

    // GetLabelSelector returns the query parameter value of the label selector

> 	GetLabelSelector() string

Reviewer: I'm not sure we need access to the raw strings in the interface now, if we authorized unrestricted when parse fails, and always sent only parsed requirements to webhooks

    // It can return an error if the field selector cannot be parsed.
    ParseLabelSelector() ([]Requirement, error)
```

### SubjectAccessReview Changes

SubjectAccessReview is use for two purposes:
1. Authorization webhook calls from the kube-apiserver to a webhook.
   This usage likely benefits from a serialization with `[]Requirement`.
2. Authorization checks from a client (often a server process using in-cluster authorization like kube-rbac-proxy)
   This usage likely benefits from a serialization that matches the query parameter.

Their needs best met with two different serialization (see user stories)

```go
type SubjectAccessReviewSpec struct {
	FieldSelector *FieldSelectorAttributes
	
	LabelSelector *LabelSelectorAttributes
}

// FieldSelectorAttributes indicates a field limited access.
// If rawSelector is empty and requirements are empty, the request is not limited.
// If rawSelector is present and requirements are empty, the rawSelector should be honored.
// If rawSelector is empty and requirements are present, the requirements should be honored
// If rawSelector is present and requirements are present, the implementation must describe which (or both) is honored.

> // Remember, the rawSelector may not match the requirements!

Reviewer 2: What is the usecase for these two not matching each other?

Author: > What is the usecase for these two not matching each other?

No use-case, but recognition that clients can place any data inside a request and not all combinations make sense.  Webhook authorizers are likely to pay the cost to reparse and validate the input since the only impacted party is the requestor.

Author: the kube-apiserver can control what it sends and what it accepts, but general webhook authors cannot control what is sent from all who have access to it.  Even if we spec them to match, a naughty client can lie and authors should be expecting to handle such a case.

Author: >This eliminates the mismatch edge and ensures webhooks don't do naive incorrect checks against the raw string
> require they match if both are set

We can say in the doc it is required.  We cannot ensure that a naughty client doesn't directly send a mismatched request to a webhook.

Reviewer: as long as kube-apiserver always transforms raw strings into parsed requirements (if they parse), and never sends a raw string to a webhook in a SAR, I'm fine with letting naughty clients and clever webhooks those naughty clients talk directly to do whatever they want

Reviewer: I would tell webhook authors to only look at requirements and authorize any SAR that doesn't include requirements as an unfiltered request

	// merge patch.
	// +optional
	// +listType=atomic
	Values []string `json:"values,omitempty" protobuf:"bytes,3,rep,name=values"`
}

```

Importantly, if old webhook authorizers do not honor these new field, they will assume the broadest possible access and fail closed.
If old in-cluster authorization does not include field and label selectors, the kube-apiserver will assume the broadest possible access and fail closed.

### User Stories (Optional)

#### As a SAR client, I want to check a request with a field or label selector

This type of usage probably finds the stringified serialization format used in the query parameters the
most convenient format to build their request with.
Providing the query parameter serialization format avoids the need for a client to grow a decently complex lexer/parser.

#### As an authorization webhook author, I want to easily consume the field and label selectors

This type of usage probably finds a serialized `[]Requirement` to be the most convenient way to consume the field and label selector.
Providing the parsed value avoids the need for every consumer to grow a decently complex lexer/parser.

> ### Notes/Constraints/Caveats (Optional)

Reviewer: some random spots to not forget to update:

* authorization webhook matchConditions, which evaluates the v1 SubjectAccessReview that _would_ be sent to the webhook https://github.com/kubernetes/kubernetes/blob/bb838fde5bb9df4becb9fd267c84759be9f5400f/staging/src/k8s.io/apiserver/pkg/authorization/cel/compile.go#L197-L205
* v1 / v1beta1 SAR translation function https://github.com/kubernetes/kubernetes/blob/bb838fde5bb9df4becb9fd267c84759be9f5400f/staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go#L472-L485
* v1 SAR construction function https://github.com/kubernetes/kubernetes/blob/bb838fde5bb9df4becb9fd267c84759be9f5400f/staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go#L198
* cache size decision https://github.com/kubernetes/kubernetes/blob/bb838fde5bb9df4becb9fd267c84759be9f5400f/staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go#L440

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

#### client provides field or label selector to kube-apiserver that does not parse

The kube-apiserver does not parse the selectors until requested.

> This means that if the authorizers somehow use string matching, they may use the rawSelector values to authorize.

Reviewer: I'd prefer not to enable this

It will be up to the REST handler to accept or reject requests for bad selectors.

> This approach allows an aggregated API server to have extended field and label selector syntax, though we strongly discourage doing so.

Reviewer: This seems like a bad thing to enable... I think a selector that doesn't parse should be authorized without the selector information (your option #3)

If the authorizer uses the parsed selectors (far more likely), the authorizer should make an explicit choice about what to do.
Options include
1. Error on unparseable selector and NoOpinion (continue to next authorizer).
2. Error on unparseable selector and Fail (do not continue to next authorizer, we know we should have authorized)
3. Attempt to authorize without the selector information.
   If the client is authorized without the selector, then Allow since they have broader permission.
   If the client is not authorized without the selector then either NoOpinion or Fail depending on intent.

> I suggest option #3 as generally the best.

Reviewer: I agree option 3 is best

> #### client provides field or label selector to kube-apiserver with improper verb

Reviewer: what about a SAR that includes selector requirements and a non-list/watch/deletecollection verb?

Consider a client that sends an Update request with a field selector on it.
The metav1.UpdateOption doesn't allow this, but imagine devious-user with an alternative library.
The `ResolveRequestInfo` method will not add field and label selectors to the `requestInfo`, so they will not appear
in the `authorization.Attributes`, so the spurious selectors are not passed to the authorizer.
This keeps authorization behavior exactly as it was previously.

#### client provides SAR where field rawSelector does not match field requirements.

> The results are undefined.  Different webhooks can choose to handle this differently.

Reviewer: What is the downside of making these mutually exclusive when submitting a SAR to kube-apiserver, and if the raw one is set, always parsing and setting `requirements` (if parse was successful) in kube-apiserver → webhook calls?

Any authorizer that gets an error from `ParseFieldSelector` or `ParseLabelSelector` may attempt to authorize without
field or label selectors since that will authorize using a wider permission (field and label selectors can only reduce access).

```go
type Attributes interface {
	// ParseLabelSelector is lazy, thread-safe, and stores the parsed result and error.
	// It can return an error if the field selector cannot be parsed.
    // Remember that field selector formats vary based on the version of the API being used!
	ParseFieldSelector() ([]Requirement, error)
  
	// ParseLabelSelector is lazy, thread-safe, and stores the parsed result and error.
    // It can return an error if the field selector cannot be parsed.
    ParseLabelSelector() ([]Requirement, error)
```

#### Future-proofing your authentication webhook for future verbs

As of 1.31, the only verbs with field and label selectors are List, Watch, and DeleteCollection.
<<[UNRESOLVED deads2k, liggitt, jpbetz, sttts: this is a future promise for the kube-apiserver behavior if we add it ]>>

> In the future, the kube-apiserver may add field and label selectors to Create, Update, and Delete.

Reviewer: ```suggestion
In the future, the kube-apiserver may add field and label selectors to Create, Update, Patch, and Delete.
```

> In the future, the kube-apiserver may add field and label selectors to Create, Update, and Delete.

Reviewer 3: If, in the future, we add field/label selectors to Create/Update/Patch/Delete, is get the only remaining unsupported CRUD operation? Should we support it in the future too?

Reviewer 3: Thinking about this more? Is an explicit selector ever needed for single resource operations? Is the actual value of the labels and fields sufficient for any authz checks that would need to be made?

Reviewer: authz doesn't have access to the body of the incoming object or the persisted content of the existing object

The query here would be like a self-attesting precondition the writer could optionally add that would be visible to authz and would have to be verified/blocking at the storage layer for the write to succeed.

The opt-in nature and the duplication with the body content is… a little weird, so I'm not sure if this is a direction we'd ever go, but defining how the precondition would be treated *if* we did is important.

Reviewer 3: OK, this explains it. I was wondering if this might be the motivation.  Including this background in the KEP about _why_ we could consider passing a field selector alongside a single resource operation that redundantly states information already in the resource seems valuable here.

<<[/UNRESOLVED]>>
* For Create, this means that the resource after all mutation is complete (finalObject) must match the field and label selector. 

> * For Update, this means that the finalNewObject and oldObject must match the field and label selector.

Reviewer: ```suggestion
* For Update/Patch, this means that the finalNewObject and oldObject must match the field and label selector.
```

* For Delete, this means that the oldObject must match the field and label selector.
 
We do not expand field and label selectors for Get, because if a client is specifying a selector, they can add a `.metadata.name`

> field selector and use a List to get equivalent functionality.

Author: @jpbetz @sttts **If** we add field and label selectors, I'd like to provide advice to the authentication webhook authors about what the semantics will be.  This is not a proposal to make those changes.

My intent is that if a field and/or label selector are specified on a mutation request, the object being mutated (both old and new) are guaranteed to match the selectors.  If the object (old or new) does not match the selector, the storage layer rejects the request.

Reviewer: Do we need to define meaning of selectors if/when included for subresources in the future?

The main question I'd have is whether a distinction is made for subresources which do and do not see the full object in question (admission defines `objectSelector` to only work for subresources which see the full object)

* examples of ones which see the full object: `pods/status`
* examples of ones which do not see the full object: `pods/{exec,attach,log,proxy,binding}`, `deployments/scale`

Author: >The main question I'd have is whether a distinction is made for subresources which do and do not see the full object in question (admission defines objectSelector to only work for subresources which see the full object)

I'm nearly in agreement.  But I think that the storage layer of deployments/scale has the entire object. Versus admission which does not have the entire object.  `store *genericregistry.Store` appears to be shared between spec, status, and scale.

Subresources that are not stored into etcd would not be able to enforce as uniformally.  We could select specific subresources that we know end up in storage like status and scale.  pods/binding and eviction would both be useful, but far narrower.

Reviewer: > if a field and/or label selector are specified on a mutation request, the object being mutated (both old and new) are guaranteed to match the selectors

I think this generalization works for subresources as well.

The storage layer must ensure the parent object matches the selectors and that the request does not change that (either because the request cannot modify labels / fields or after applying the mutations in the request it still matches the selectors).

If the storage layer cannot verify the parent object matches the selectors because it doesn't have visibility, the request would be rejected.

(for interested bystanders, this is all *very* hypothetical future oriented ... there is no current plan to add selectors to mutating requests or to subresources ... we just want the meaning to be well-defined so an authorizer making decisions today knows what it could be allowing in the future)

### SubjectAccessReview Changes

SubjectAccessReview is used for two purposes:
1. Authorization webhook calls from the kube-apiserver to a webhook.
   This usage likely benefits from a serialization with `[]Requirement`.
2. Authorization checks from a client (often a server process using in-cluster authorization like kube-rbac-proxy)
   This usage likely benefits from a serialization that matches the query parameter.

Their needs best met with two different serialization (see user stories)

```go
type SubjectAccessReviewSpec struct {
	FieldSelector *FieldSelectorAttributes
	

> 	LabelSelector *LabelSelectorAttributes

Reviewer: would these go under `ResourceAttributes` rather than directly under the `SubjectAccessReviewSpec`?

}

// FieldSelectorAttributes indicates a field limited access.
// For webhooks:
// The kube-apiserver will never send a request with rawSelector set, but we cannot control what other clients directly send.
// If rawSelector is empty and requirements are empty, the request is not limited.
// If rawSelector is present and requirements are empty, the request is not limited.
// If rawSelector is empty and requirements are present, the requirements should be honored
// If rawSelector is present and requirements are present, the request is invalid.
// For the kube-apiserver:
// If rawSelector is empty and requirements are empty, the request is not limited.
// If rawSelector is present and requirements are empty, the rawSelector will be parsed and limited if the parsing succeeds.
// If rawSelector is empty and requirements are present, the requirements should be honored

> // If rawSelector is present and requirements are present, the request is invalid.

Reviewer: could these be simplified to:

```
* If `rawSelector` and `requirements` are both present, the request is invalid.
* For kube-apiserver: if `rawSelector` is present, the rawSelector will be parsed and honored if the parsing succeeds and the verb supports selectors.
* For kube-apiserver and webhooks: if requirements are present, the requirements should be honored if the verb supports selectors.
```

Author: > the requirements should be honored if the verb supports selectors.

How and why would an authorizer make a decision about whether the client that called it supports selectors for a particular verb?  If we make this a webhook responsibility, would this mean that the webhook needs to predict when the kube-apiserver will honor selectors for verb and subresource tuples?

Author: I find the fully separation much easier to find what I'm looking for.  Your feel strongly about simplifying?  Do you want the change the to webhook semantics from what I proposed?  I don't see how it's possible.

type FieldSelectorAttributes struct {

> 	// rawSelector is the serialization of a field selector that would be included in a List query parameter.

Reviewer: ```suggestion
	// rawSelector is the serialization of a field selector that would be included in a query parameter.
```

> 	// Webhook implementations must handle malformed rawSelectors.

Reviewer: > 	// Webhook implementations must handle malformed rawSelectors.

Must they? A lot of the docs seem leftover from the first draft where the fields were not mutually exclusive

I'd encourage general webhook authors to:
1. ensure rawSelector and requirements are not both set
2. consider the `requirements` field if set
3. not try to parse or consider the `rawSelector` field if set

similar questions on the other docs and other selector type

Author: > Must they? A lot of the docs seem leftover from the first draft where the fields were not mutually exclusive
> 
> I'd encourage general webhook authors to:

how would you like to phrase: "clients can send you data that is invalid that a kube-apiserver will never send you.  Think about it advance and don't crash"?

Reviewer: I think this covers it?

> I'd encourage general webhook authors to:
> 
> * ensure rawSelector and requirements are not both set
> * consider the requirements field if set
> * not try to parse or consider the rawSelector field if set

Author: ok, willing to update.

Trying to decide how much carries to actual API documentation that appears on the SAR endpoint.  I haven't checked docs to see if we have a separate place yet.

	RawSelector string

> 	// requirements is the parsed interpretation of the rawSelector.

Reviewer: ```suggestion
	// requirements is the parsed interpretation of a field selector.
```

// If rawSelector is empty and requirements are present, the requirements should be honored
// If rawSelector is present and requirements are present, the request is invalid.
type LabelSelectorAttributes struct {
	// rawSelector is the serialization of a field selector that would be included in a List query parameter.
	// Webhook implementations must handle malformed rawSelectors.
	RawSelector string

    // requirements is the parsed interpretation of the rawSelector.
    // All requirements must be met for a resource instance to match the selector.
    // The two should match, but recall that clients can craft any kind of request.
    // Webhook implementations must handle mismatched rawSelector and requirements, but how to handle them is up to the webhook.
	Requirements []metav1.LabelSelectorRequirement
}

type FieldSelectorRequirement struct {
	// key is the label key that the selector applies to.
	Key string `json:"key" protobuf:"bytes,1,opt,name=key"`
	// operator represents a key's relationship to a set of values.
	// Valid operators are In, NotIn, Exists, DoesNotExist

> 	Operator LabelSelectorOperator `json:"operator" protobuf:"bytes,2,opt,name=operator,casttype=LabelSelectorOperator"`

Reviewer: since this is essentially an enum field, should describe how clients should handle unrecognized operators if we ever want the ability to expand this in the future

easiest way is to ignore entries with unrecognized operators and assume they don't limit the request

A single-item List or Watch request is still a list as normal (including selectors), but also includes a name.

### Authorization Attributes changes

The authorization attributes have easy access to the query parameter field and label selectors.
To avoid confusion, field and label selectors will not be included in authorization attributes for verbs where the field
selector has no semantic meaning.
In practice this means that (for now), only List, Watch, and DeleteCollection have field and label selectors.
SubjectAccessReviews submitted to the kube-apiserver with verbs that do not honor the selectors will have the field and
label selector attributes removed prior to checking authorization.
Webhook authors: remember that the list of verbs accepting field and label selectors may change over time.
If the kube-apiserver sends the FieldSelector or LabelSelector to a webhook, the kube-apiserver intends to honor the selector attributes.
Any authorizer that gets an error from `ParseFieldSelector` or `ParseLabelSelector` may attempt to authorize without
field or label selectors since that will authorize using a wider permission (field and label selectors can only reduce access).

```go

> type Attributes interface {

Author: > Are we concerned that if we add these two methods to the `authorizer.Attributes` interface now, implementers of the interface won't satisfy it on a go.mod update without manually adding those methods? We don't make promises AFAIK for stuff like that, but just asking.

I'm ok breaking people that integrated at this level.  They should be aware of the semantic change and this is an easy way to force it.

	// It can return an error if the field selector cannot be parsed.
    // Remember that field selector formats vary based on the version of the API being used!
	ParseFieldSelector() ([]Requirement, error)
  
	// ParseLabelSelector is lazy, thread-safe, and stores the parsed result and error.
    // It can return an error if the field selector cannot be parsed.
    ParseLabelSelector() ([]Requirement, error)
```

#### Future-proofing your authentication webhook for future verbs

As of 1.31, the only verbs with field and label selectors are List, Watch, and DeleteCollection.
<<[UNRESOLVED deads2k, liggitt, jpbetz, sttts: this is a future promise for the kube-apiserver behavior if we add it ]>>
In the future, the kube-apiserver may add field and label selectors to Create, Update, Patch, and Delete.
<<[/UNRESOLVED]>>
* For Create, this means that the resource after all mutation is complete (finalObject) must match the field and label selector. 
* For Update/Patch, this means that the finalNewObject and oldObject must match the field and label selector.
* For Delete, this means that the oldObject must match the field and label selector.
 

> We do not expand field and label selectors for Get, because if a client is specifying a selector, they can add a `.metadata.name`

Author: clarified

field selector and use a List to get equivalent functionality.

### SubjectAccessReview Changes

SubjectAccessReview is used for two purposes:
1. Authorization webhook calls from the kube-apiserver to a webhook.
   This usage likely benefits from a serialization with `[]Requirement`.

> 2. Authorization checks from a client (often a server process using in-cluster authorization like kube-rbac-proxy)

Author: > Is parsing the URI into requirements really that bad that we have to add two modes, and make the matrix of parsing the SAR much much harder? I would think the opposite

It's non-trivial and not all webhook implementations are in go, so providing a golang library (note that we already include parsers in apimachinery) falls well short.

SAR to the kube-apiserver will parse the RawSelector for convenience.

> 2. Authorization checks from a client (often a server process using in-cluster authorization like kube-rbac-proxy)

Reviewer 4: @luxas https://github.com/kubernetes/enhancements/pull/4600#discussion_r1595580964

> 2. Authorization checks from a client (often a server process using in-cluster authorization like kube-rbac-proxy)

Reviewer 5: Fair enough wrt. impls in other languages.

> SAR to the kube-apiserver will parse the RawSelector for convenience.

Ok this is good. At least minimizing the complexity webhooks have to deal with is good.

@enj thanks for giving heads up on that CVE.

// For webhooks:
// The kube-apiserver will never send a request with rawSelector set, but we cannot control what other clients directly send.
// If rawSelector is empty and requirements are empty, the request is not limited.
// If rawSelector is present and requirements are empty, the request is not limited.
// If rawSelector is empty and requirements are present, the requirements should be honored
// If rawSelector is present and requirements are present, the request is invalid.
// For the kube-apiserver:
// If rawSelector is empty and requirements are empty, the request is not limited.
// If rawSelector is present and requirements are empty, the rawSelector will be parsed and limited if the parsing succeeds.
// If rawSelector is empty and requirements are present, the requirements should be honored
// If rawSelector is present and requirements are present, the request is invalid.
type LabelSelectorAttributes struct {
	// rawSelector is the serialization of a field selector that would be included in a query parameter.
    // Webhook implementations are encouraged to ignore rawSelector.
	// The kube-apiserver's SubjectAccessReview will parse the rawSelector. 

> 	RawSelector string

Reviewer 5: Talking about RawSelector, at the moment we only support Equals/NotEquals/Exists/NotExists in the url encoding, right? As we're looking into this, how do we feel about adding In/NotIn semantics like `labelSelector=foo=a|b|c`? I guess it's kind of impossible at this point, because we don't limit the content set on label values, so we cannot choose a character that can act as the separator.

Another thing I was thinking is `labelSelector=foo=a,foo=b,foo=c`, but that ends up being three different requirements, which ANDed together becomes an unsatisfiable requirements set (see the logic below).

I don't know, but it feels like if we start expanding the scope of these selectors, that's a practical issue that instantly arises for consumers.

Author: Currently every discrete requirement is AND'd when evaluated server-side.  I don't see a need to change that behavior.

Reviewer: > Currently every discrete requirement is AND'd when evaluated server-side. I don't see a need to change that behavior.

Agreed. In fact if we do change it from AND we break the ability to expand operators in the future safely.

With AND semantics, a webhook can check for restrictions they understand, short-circuit when they find a limiting one they are satisfied by, ignore ones with operators they don't recognize and treat them as non-limiting, and fail in a safe direction in the presence of unrecognized operators.

> it feels like if we start expanding the scope of these selectors, that's a practical issue that instantly arises for consumers.

Given selectors are already part of our API, both in string and structured form, I don't think plumbing the current semantics into authz really expands their scope, especially if we give explicit instructions for how to treat unrecognized operators that makes it safe to accommodate operator expansion in the future.

Reviewer 5: Sorry if my comment was unclear. Not saying we should move away from ANDed requirements, that should definitely be the case.

I was talking about the fact that within a single requirement, the structured YAML/JSON requirements are more expressible than their string encoding, as the `In` and `NotIn` operators do not convert to string form. So just pointing out that one hard part about using the current selector API in string form is that you cannot say "watch objects with labels/fields with key `foo` equals `a` or `b`.

So I was thinking if there are ways to start supporting that in the future, given that the API server is the canonical place for parsing these queries now, but I'm afraid if it isn't possible to start treating some characters as the "values separator" as a value could be almost anything.

Author: > I'm afraid if it isn't possible to start treating some characters as the "values separator" as a value could be almost anything.

This came up separately when discussion potential future changes to field selectors.  The problem is recognized, but the construction used here in []Requirements (recommended for webhook usage) gives us flexibility to change how field selectors are specified.

    // requirements is the parsed interpretation of a label selector.
    // All requirements must be met for a resource instance to match the selector.
    // Webhook implementations should handle requirements, but how to handle them is up to the webhook.
    // Since requirements can only limit the request, it is safe to authorize as unlimited request if the requirements
    // are not understood.
	Requirements []metav1.LabelSelectorRequirement
}

type FieldSelectorRequirement struct {
	// key is the label key that the selector applies to.

> 	Key string `json:"key" protobuf:"bytes,1,opt,name=key"`

Author: > What format is this?

Arbitrary string with no semantic meaning attached.  That's the meaning today and I don't propose to change it. CRs are slightly more strict.

	// operator represents a key's relationship to a set of values.
	// Valid operators are In, NotIn, Exists, DoesNotExist
	// The list of operators may grow in the future.
	// Webhook authors are encouraged to ignore unrecognized operators and assume they don't limit the request.

> 	// The semantics of "all requirements are AND'd will not change, so other requirements can continue to be enforced.

Reviewer 5: I believe we should provide a function to "simplify" the requirements into a "canonical representation". For example, if there are multiple rules for the same key, one can apply the following rules:

1. Simplify Equals/NotEquals to In/NotIn (this already happens in the rawSelector parsing layer essentially)
Equals `foo` => In `[foo]`
NotEquals `foo` => NotIn `[foo]`

2. Simplify In/NotIn separately

In `[a,b]` && In `[b,c]` => In `[b]` (intersection)
NotIn `[a,b]` && NotIn `[b,c]` => NotIn `[a,b,c]` (union)

By this stage, we have at most one In and one NotIn requirement per key (plus any possible Exists/NotExist, handled later)

3. Simplify In/NotIn together

In `[a,b]` && NotIn `[b,c]` => In `[a]` (subtraction)

Note: this must be done last (as far as I can see), as the previous union/intersection operations are commutative, but subtraction is not.

In the end of this process we have at most one requirement with operator In or NotIn (not both), plus any possible Exists/NotExist, handled later.

4. Simplify Exists/NotExists

Exists && Exists => Exists
NotExists && NotExists => NotExists
Exists && NotExists => ∅ (a label cannot both be defined and not for the same object, unsatisfiable)

At this point, we have at most one (In or NotIn) and one (Exists or NotExists).
Of course, we can short-circuit earlier if we get an empty set before, as then the requirements are unsatisfiable by any object.

5. Combine one (In or NotIn) and one (Exists or NotExists)

In `[a,b]` && Exists => In `[a,b]` (if foo=a or foo=b is true, foo does exist)
NotIn `[a,b]` && Exists => NotIn `[a,b]` && Exists (no simplification possible)
In `[a,b]` && NotExists => ∅ (if the label must both be set and not set, req is unsatisfiable)
NotIn `[a,b]` && NotExists => NotExists (a label not existing certainly is not set to a given value)

At this point, of all possible requirements, we have at most the following requirements for each key:
In `[a,b]`
NotIn `[a,b]`
Exists
NotExists
∅ (unsatisfiable, which makes the whole requirements list for any object unsatisfiable)
NotIn `[a,b]` && Exists

This "simplified/parsed" requirements set could provide a dedicated interface, which can answer e.g. `ObjectMatches(obj)`, `Unsatisfiable()`, `Equal(otherRequirements)`, `GreaterThan(otherRequirements`, etc., such that one can semantically compare varying requirements sets with each other (a use case that will be very much needed).

Author: > I believe we should provide a function to "simplify" the requirements into a "canonical representation". For example, if there are multiple rules for the same key, one can apply the following rules:

I don't see a reason to do this at the moment.

Reviewer 5: Yeah. Not needed for this KEP indeed, but I have some use-cases where this would be useful, so I can experiment with that outside of core indeed.

	Operator LabelSelectorOperator `json:"operator" protobuf:"bytes,2,opt,name=operator,casttype=LabelSelectorOperator"`
	// values is an array of string values. If the operator is In or NotIn,
	// the values array must be non-empty. If the operator is Exists or DoesNotExist,
	// the values array must be empty. This array is replaced during a strategic
	// merge patch.
	// +optional
	// +listType=atomic
	Values []string `json:"values,omitempty" protobuf:"bytes,3,rep,name=values"`
}

```

Importantly, if old webhook authorizers do not honor these new field, they will assume the broadest possible access and fail closed.
If old in-cluster authorization does not include field and label selectors, the kube-apiserver will assume the broadest possible access and fail closed.

### User Stories (Optional)

#### As a SAR client, I want to check a request with a field or label selector

This type of usage probably finds the stringified serialization format used in the query parameters the
most convenient format to build their request with.

> Providing the query parameter serialization format avoids the need for a client to grow a decently complex lexer/parser.

Author: > Someone always have to parse the rawQuery at the end of the day,

In this proposal, that someone is the kube-apiserver (or aggregated apiserver).

> Providing the query parameter serialization format avoids the need for a client to grow a decently complex lexer/parser.

Reviewer: > I'd like to do that in the place where it doesn't blow up the complexity for others, i.e. in this SAR client.

being able to pass the string form to kube-apiserver to parse also avoids parsing complexity for SAR clients

>  (with well-defined Go libraries to handle this easily, of course).

pushing parsing to clients via versioned / compiled-in libraries limits our ability to evolve to support new requirement operators in the future, and doesn't help clients implemented in languages other than go

> In this proposal, that someone is the kube-apiserver (or aggregated apiserver).

I agree with this approach

> Providing the query parameter serialization format avoids the need for a client to grow a decently complex lexer/parser.

Reviewer 5: > being able to pass the string form to kube-apiserver to parse also avoids parsing complexity for SAR clients

Right, I see, I was thinking about this from the webhook perspective. If API server never sends a SAR webhook with the rawQuery set, then all is good 👍 thanks for the clarification

[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

The authorization attributes will be extended to include field selectors and label selectors from
List, Watch, and DeleteCollection.
This will allow authorizers to use these selectors when making an authorization decision.

## Motivation

Security for per-node workloads could be improved by exposing field and label selectors to authorizers.
Adding them as authorization attributes allows the development of new kinds of authorizers that
leverage this information to provide security.
In particular, it enables out-of-tree authorizers to experiment with ways to express restrictions based on field and label selectors.

### Goals

* Add field and label selectors to authorization attributes for List, Watch, and DeleteCollection verbs.
* Add field and label selectors to webhook authorization types.
* Add field and label selectors to SelfSubjectAccessReview (SSAR), SubjectAccessReview (SAR), and LocalSubjectAccessReview.
* Update node authorizer to restrict on nodeName field selector.

### Non-Goals

* Create a generic in-tree authorizer that manages field or label selectors.

> * Expand the audit surface area, since requestURI is already included

Author: > How do I, as a cluster admin/operator, know exactly what SAR a request was authorized for?

given a request URI, you know the field and label selectors used.

* Expand the admission surface area (admission.Attributes, AdmissionReview, CEL authorizer implementation available to admission) 

>   since admission verbs don't support field/label selectors

Author: > What if a CEL authz call wants to check list with selectors?

Interesting thought.  I'll have a think on it.

>   since admission verbs don't support field/label selectors

Author: I agree this should be addressed and that it's an API change, so it must be done during the alpha stage.  As I recall, CEL library modifications take two releases to roll out.

	ParseFieldSelector() ([]Requirement, error)
  
	// ParseLabelSelector is lazy, thread-safe, and stores the parsed result and error.
    // It can return an error if the field selector cannot be parsed.
    ParseLabelSelector() ([]Requirement, error)
```

#### Future-proofing your authentication webhook for future verbs

As of 1.31, the only verbs with field and label selectors are List, Watch, and DeleteCollection.
<<[UNRESOLVED deads2k, liggitt, jpbetz, sttts: this is a future promise for the kube-apiserver behavior if we add it ]>>
In the future, the kube-apiserver may add field and label selectors to Create, Update, Patch, and Delete.
<<[/UNRESOLVED]>>
* For Create, this means that the resource after all mutation is complete (finalObject) must match the field and label selector. 
* For Update/Patch, this means that the finalNewObject and oldObject must match the field and label selector.
* For Delete, this means that the oldObject must match the field and label selector.
 
We do not allow field and label selectors for Get, because if a client is specifying a selector, they can add a `.metadata.name`

> field selector and use a List to get equivalent functionality.

Author: > Are we sure this is good enough?

I can't think of what else Get is beyond a single-item list.

> field selector and use a List to get equivalent functionality.

Reviewer 6: Lots of clients are using GET today though. I think there's an assumption here that clients would have to change to inject label selectors anyway, so they might as well switch to single-element LIST, but that might not always be true. I'd also be curious if there's any performance difference between single-element LIST with field selector and GET.

> field selector and use a List to get equivalent functionality.

Reviewer 5: > I'd also be curious if there's any performance difference between single-element LIST with field selector and GET.

Yeah this would be interesting to benchmark.

> field selector and use a List to get equivalent functionality.

Author: Added an explanation of what field and label selectors would do with a GET, but I'm not 100% sold on whether we'd actually add it.

- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

The authorization attributes will be extended to include field selectors and label selectors from

> List, Watch, and DeleteCollection.

Reviewer 6: GET doesn't support label selectors today... but something in that direction would be _really_ nice for full coverage of read authorization.

Reviewer 6: You can technically handle UPDATE, PATCH, DELETE in admission control by checking the label on the old object, but having label selectors supported for any verb and all possible to check in authorization policy  would probably be desirable at some point for writing consistent policy...

Author: Added how it could be done, but whether we do it is a debate for another day.

<<[UNRESOLVED deads2k, liggitt, jpbetz, sttts: this is a future promise for the kube-apiserver behavior if we add it ]>>
In the future, the kube-apiserver may add field and label selectors to Create, Update, Patch, and Delete.
<<[/UNRESOLVED]>>
* For Create, this means that the resource after all mutation is complete (finalObject) must match the field and label selector. 
* For Update/Patch, this means that the finalNewObject and oldObject must match the field and label selector.
* For Delete, this means that the oldObject must match the field and label selector.
 
We do not allow field and label selectors for Get, because if a client is specifying a selector, they can add a `.metadata.name`
field selector and use a List to get equivalent functionality.

### SubjectAccessReview Changes

SubjectAccessReview is used for two purposes:
1. Authorization webhook calls from the kube-apiserver to a webhook.
   This usage likely benefits from a serialization with `[]Requirement`.
2. Authorization checks from a client (often a server process using in-cluster authorization like kube-rbac-proxy)
   This usage likely benefits from a serialization that matches the query parameter.

> Their needs best met with two different serialization (see user stories)

Reviewer: ```suggestion
Their needs are best met with two different serialization (see user stories)
```

	ResourceAttributes *ResourceAttributes
}

type ResourceAttributes struct {
    // If the SubjectAccessReview specified verb,resource tuple does not honor field selection, the fieldSelector is ignored.
	FieldSelector *FieldSelectorAttributes

    // If the SubjectAccessReview specified verb,resource tuple does not honor label selection, the labelSelector is ignored.
	LabelSelector *LabelSelectorAttributes
}

// FieldSelectorAttributes indicates a field limited access.
// For webhooks:
// The kube-apiserver will never send a request with rawSelector set, but we cannot control what other clients directly send.
// * If rawSelector is empty and requirements are empty, the request is not limited.
// * If rawSelector is present and requirements are empty, the request is not limited.
// * If rawSelector is empty and requirements are present, the requirements should be honored
// * If rawSelector is present and requirements are present, the request is invalid.
// Webhook authors are encouraged to
// * ensure rawSelector and requirements are not both set
// * consider the requirements field if set
// * not try to parse or consider the rawSelector field if set.

> //   This avoid this is to avoid another CVE-2022-2880 (i.e. getting different systems to agree on how exactly to parse

Reviewer: ```suggestion
//   This is to avoid another CVE-2022-2880 (i.e. getting different systems to agree on how exactly to parse
```

//   a query is not something we want), see https://www.oxeye.io/resources/golang-parameter-smuggling-attack for more details.
// For the kube-apiserver:
// * If rawSelector is empty and requirements are empty, the request is not limited.
// * If rawSelector is present and requirements are empty, the rawSelector will be parsed and limited if the parsing succeeds.

> // * If rawSelector is empty and requirements are present, the requirements should be honored

Reviewer: from https://github.com/kubernetes/enhancements/pull/4600/files#r1593019928

> client -> kube-apiserver SAR endpoint: if the kube-apiserver would not honor the selectors, they are removed prior to calling the authorizer.

for the "rawSelector is present" and "requirements are present" cases, is there a verb constraint as well? The verb must be one kube-apiserver would honor the selector for in order for the request to be limited?

(same question on label selector docs)

Author: > for the "rawSelector is present" and "requirements are present" cases, is there a verb constraint as well? The verb must be one kube-apiserver would honor the selector for in order for the request to be limited?

Doing so would preclude a N+1 aggregated apiserver that allows additional verbs from using in-cluster authorization if the kube-apiserver doesn't also know about that verb.  Allowing any verb keeps this possible and only impacts the client.

Reviewer: I guess I don't understand how to reconcile that with the text at https://github.com/kubernetes/enhancements/pull/4600/files#r1593019928

Author: > Does kube-apiserver honor selector requirements in a SAR for a verb that doesn't support selectors or not?

When receiving a request, the request info resolver only adds field and label selector information to authorization attributes when the kube-apiserver honors those selectors for the verb being used.

When handling a SAR, no verb interpretation is done prior to adding the field and label selectors.

This allows skew in the cluster while ensuring the kube-apiserver request handling only considers field and label selectors for certain verbs.

Reviewer: > When receiving a request, the request info resolver only adds field and label selector information to authorization attributes when the kube-apiserver honors those selectors for the verb being used.
>
> When handling a SAR, no verb interpretation is done prior to adding the field and label selectors.
>
> This allows skew in the cluster while ensuring the kube-apiserver request handling only considers field and label selectors for certain verbs.

Ok, I think I agree with that. Sweep the KEP / API docs to make that consistent and clear.

For example, the text in the KEP still says "SubjectAccessReviews submitted to the kube-apiserver with verbs that do not honor the selectors will have the field and label selector attributes removed prior to checking authorization." which I think is directly opposed to "When handling a SAR, no verb interpretation is done prior to adding the field and label selectors"

Author: Updated.  Caught a couple others too.

// * If rawSelector is empty and requirements are empty, the request is not limited.
// * If rawSelector is present and requirements are empty, the rawSelector will be parsed and limited if the parsing succeeds.
// * If rawSelector is empty and requirements are present, the requirements should be honored
// * If rawSelector is present and requirements are present, the request is invalid.
type LabelSelectorAttributes struct {
	// rawSelector is the serialization of a field selector that would be included in a query parameter.
    // Webhook implementations are encouraged to ignore rawSelector.
	// The kube-apiserver's SubjectAccessReview will parse the rawSelector. 
	RawSelector string

    // requirements is the parsed interpretation of a label selector.
    // All requirements must be met for a resource instance to match the selector.
    // Webhook implementations should handle requirements, but how to handle them is up to the webhook.
    // Since requirements can only limit the request, it is safe to authorize as unlimited request if the requirements
    // are not understood.
	Requirements []metav1.LabelSelectorRequirement
}

type FieldSelectorRequirement struct {

> 	// key is the label key that the selector applies to.

Reviewer: ```suggestion
	// key is the field selector key that the requirement applies to.
```

	Key string `json:"key" protobuf:"bytes,1,opt,name=key"`
	// operator represents a key's relationship to a set of values.
	// Valid operators are In, NotIn, Exists, DoesNotExist
	// The list of operators may grow in the future.
	// Webhook authors are encouraged to ignore unrecognized operators and assume they don't limit the request.
	// The semantics of "all requirements are AND'd will not change, so other requirements can continue to be enforced.
	Operator LabelSelectorOperator `json:"operator" protobuf:"bytes,2,opt,name=operator,casttype=LabelSelectorOperator"`
	// values is an array of string values. If the operator is In or NotIn,
	// the values array must be non-empty. If the operator is Exists or DoesNotExist,
	// the values array must be empty. This array is replaced during a strategic

> 	// merge patch.

Reviewer: drop the "This array is replaced during a strategic merge patch." bit?

	// +optional
	// +listType=atomic
	Values []string `json:"values,omitempty" protobuf:"bytes,3,rep,name=values"`
}

```

> Importantly, if old webhook authorizers do not honor these new field, they will assume the broadest possible access and fail closed.

Reviewer: ```suggestion
Importantly, if old webhook authorizers do not honor these new fields, they will assume the broadest possible access and fail closed.
```

If old in-cluster authorization does not include field and label selectors, the kube-apiserver will assume the broadest possible access and fail closed.

### Node Authorizer Changes

> The node authorizer will `node_authorizer.go` to check the `spec.nodeName` for List requests to ensure it matches the

Reviewer: ```suggestion
The node authorizer will be modified to only authorize node clients to `list` and `watch` pods with fieldSelectors containing `spec.nodeName=$nodeName`
```

> `nodeName` from its username (`system:node:<nodeName>`).

Reviewer: The node authorizer will be modified to authorize pod `get` requests based on the graph.

> Then we will remove the 

Reviewer: We'll actually keep it in the bootstrap policy so clusters using clusters binding to system:nodes and not using NodeAuthorizer will continue to work, and just make get / list / watch pods requests ones the node authorizer authorizes using the graph and selectors rather than falling through to the bootstrap rules.

```go
		// TODO: restrict to pods scheduled on the bound node once field selectors are supported by list/watch authorization
		rbacv1helpers.NewRule(Read...).Groups(legacyGroup).Resources("pods").RuleOrDie(),
```
from `bootstrappolicy/policy.go`.

### CEL Authorizer Changes

While admission isn't supported on List, Watch, or DeleteCollection, it is reasonable to expect that secondary authorization
checks may desire to use those verbs and leverage the field and label selector capabilities.
To support this we will two congruent options similar to
```go
	"fieldSelector": {
		cel.MemberOverload("resourcecheck_fieldselector", []*cel.Type{ResourceCheckType, cel.StringType}, ResourceCheckType,
			cel.BinaryBinding(resourceCheckName))},
    }
```
This will allow usage like `authorizer.group('').resource('pods').fieldSelector('spec.nodeName=foo').check('list').allowed()`.
The parsing will happen during the call to `allowed` where we track errors and have means of handling them already.

> The `cel/library` authorizer will force errors when `fieldSelector` or `labelSelector` are used for verbs that the kube-apiserver

Reviewer: should a `fieldSelector` or `labelSelector` call that fails parsing error, or should it drop the selector to be non-limiting (like the SubjectAccessReview API)?

should use of `fieldSelector` or `labelSelector` in conjunction with a verb that doesn't currently honor selectors error, or should it drop the selector to be non-limiting (like the SubjectAccessReview API)?

Author: > should a fieldSelector or labelSelector call that fails parsing error, or should it drop the selector to be non-limiting (like the SubjectAccessReview API)?

This seems to make it very difficult to discovery VAP (or binding) bugs.  While I can see a case that leniency makes more successes, I'd try to make it easy to discover bugs and error.

> should use of fieldSelector or labelSelector in conjunction with a verb that doesn't currently honor selectors error, or should it drop the selector to be non-limiting (like the SubjectAccessReview API)?

Sure, updated.

Reviewer: > While I can see a case that leniency makes more successes, I'd try to make it easy to discover bugs and error.

Hmm... if the selector was a hard-coded string, erroring seems useful ("your syntax has a bug").

If the selector is using a field from the object as a label, I guess I'd want an invalid selector due to weirdness from the object field to just skip the tighter scoping I might have gotten from the selector.

I'm fuzzy on whether it's possible to write a CEL expression defensively to tolerate an error... I don't think it is. How would you recommend someone scope with a selector if possible but fall back to unscoped if a valid selector wasn't possible to construct.

Reviewer: also, it feels weird for the CEL API to error when SubjectAccessReview doesn't for effectively the same field (rawSelector)

Author: > How would you recommend someone scope with a selector if possible but fall back to unscoped if a valid selector wasn't possible to construct.

convincing when combined with user input.  updated.
